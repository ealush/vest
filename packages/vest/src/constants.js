/**
 * @type {String} Version number derived from current tag.
 */
export const VERSION = VEST_VERSION;
