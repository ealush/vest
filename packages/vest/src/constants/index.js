export const VERSION = VEST_VERSION;
export const OPERATION_MODE_STATELESS = 'STATELESS';
export const OPERATION_MODE_STATEFUL = 'STATEFUL';
