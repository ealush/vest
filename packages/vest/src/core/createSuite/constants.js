export const SUITE_INIT_ERROR = 'Suite initialization error.';
