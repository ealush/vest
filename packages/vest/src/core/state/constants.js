export const KEY_STATE = '$vestState';
export const KEY_SUITES = '_suites';
export const KEY_CANCELED = '_canceled';
