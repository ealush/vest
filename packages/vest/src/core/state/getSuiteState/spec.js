import { getState } from '..';
import runRegisterSuite from '../../../../testUtils/runRegisterSuite';
import { SYMBOL_SUITES } from '../symbols';
import getSuiteState from '.';

const suiteName = 'suite_1';
const suiteId = 'suiteId_1';
const suiteId_2 = 'suiteId_3';

describe('getSuiteState', () => {
  it('Should return current suite state by key', () => {
    runRegisterSuite({ name: suiteName, suiteId });
    runRegisterSuite({ name: suiteName, suiteId: suiteId_2 });
    const [currentState] = getState(SYMBOL_SUITES)[suiteId];
    expect(getSuiteState(suiteId)).toBe(currentState);
  });
});
