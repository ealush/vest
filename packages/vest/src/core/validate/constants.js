/**
 * @type {String}
 */
export const SUITE_INIT_ERROR = "Suite initialization error.";
