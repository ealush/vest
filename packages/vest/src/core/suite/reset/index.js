import { OPERATION_MODE_STATEFUL } from '../../../constants';
import throwError from '../../../lib/throwError';
import context from '../../context';
import getState from '../getState';
import register from '../register';
import remove from '../remove';

/**
 * Resets suite to its initial state
 * @param {String} suiteId
 */
const reset = suiteId => {
  if (!suiteId) {
    throwError('`vest.reset` must be called with suiteId.');
  }

  let name = suiteId;
  try {
    name = getState(suiteId).name;
    remove(suiteId);
  } catch {
    /* */
  }

  context.run(
    { name, suiteId, operationMode: OPERATION_MODE_STATEFUL },
    register
  );
};

export default reset;
