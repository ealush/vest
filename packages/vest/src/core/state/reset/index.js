import { OPERATION_MODE_STATEFUL } from '../../../constants';
import runWithContext from '../../../lib/runWithContext';
import throwError from '../../../lib/throwError';
import getSuiteState from '../getSuiteState';
import registerSuite from '../registerSuite';
import remove from '../remove';

/**
 * Resets suite to its initial state
 * @param {String} suiteId
 */
const reset = suiteId => {
  if (!suiteId) {
    throwError('`vest.reset` must be called with suiteId.');
  }

  const { name } = getSuiteState(suiteId);
  remove(suiteId);

  runWithContext(
    { name, suiteId, operationMode: OPERATION_MODE_STATEFUL },
    registerSuite
  );
};

export default reset;
