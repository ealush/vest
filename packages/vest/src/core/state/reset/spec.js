import { getState, getSuites } from '..';
import resetState from '../../../../testUtils/resetState';
import runRegisterSuite from '../../../../testUtils/runRegisterSuite';
import runSpec from '../../../../testUtils/runSpec';
import testDummy from '../../../../testUtils/testDummy';
import vest from '../../../index';
import copy from '../../../lib/copy';
import { KEY_CANCELED } from '../constants';
import getSuiteState from '../getSuiteState';
import reset from '.';

const suiteId = 'suite_id';
let state;

const spec = _vest => {
  let _reset;

  if (!_vest) {
    _reset = reset;
    _vest = vest;
  } else {
    _reset = _vest.reset;
  }

  const createSuite = () => {
    const vest = _vest;
    return vest.create(suiteId, skip => {
      vest.skip(skip);

      testDummy(vest).failingAsync('field_1', { time: 250 });

      testDummy(vest).failingAsync('field_2', { time: 250 });

      testDummy(vest).failingAsync('field_3', { time: 250 });

      testDummy(vest).failingAsync('field_4', { time: 250 });
    });
  };

  beforeAll(() => {
    resetState();
  });
  describe('When invoked without suiteId', () => {
    it('Should throw', () => {
      expect(() => reset()).toThrow('`vest.reset` must be called with suiteId');
    });
  });

  describe('When suite does not exist in state', () => {
    beforeEach(() => {
      runRegisterSuite({ name: suiteId, suiteId });
      state = copy(getState());
    });

    it('Should throw an error', () => {
      expect(state).toEqual(getState());
      expect(() => _reset('nonexistent_suite')).toThrow(Error);
    });
  });

  describe('When suite exists', () => {
    let validate;

    beforeEach(() => {
      validate = createSuite();
      validate('field_2');
      validate(['field_1', 'field_3']);
    });

    test.skipOnWatch(
      'Sanity - making sure everything works as it should',
      () => {
        const suiteState = getSuiteState(suiteId);
        expect(suiteState.lagging).toHaveLength(2);
        expect(suiteState.pending).toHaveLength(2);
        expect(suiteState.lagging[0].fieldName).toBe('field_1');
        expect(suiteState.pending[0].fieldName).toBe('field_2');
        expect(suiteState.lagging[1].fieldName).toBe('field_3');
        expect(suiteState.pending[1].fieldName).toBe('field_4');
      }
    );

    it('Should remove suite from state', () => {
      expect(getSuites()).toHaveProperty(suiteId);
      expect(() => getSuiteState(suiteId)).not.toThrow();
      _reset(suiteId);
      expect(() => getSuiteState(suiteId)).toThrow();
      expect(getSuites()).not.toHaveProperty(suiteId);
    });

    it('Should set all pending and lagging tests as canceled', () => {
      const previouslyCanceled = getState(KEY_CANCELED);
      const state = getSuiteState(suiteId);
      const allCanceled = [...state.lagging, ...state.pending].reduce(
        (canceled, { id }) => Object.assign(canceled, { [id]: true }),
        { ...previouslyCanceled }
      );
      expect(getState(KEY_CANCELED)).not.toEqual(allCanceled);
      _reset(suiteId);
      expect(getState(KEY_CANCELED)).toEqual(allCanceled);
    });
  });
};

// This tests the module when consumed through vest.reset();
runSpec(spec);

// This tests the module directly
spec();
