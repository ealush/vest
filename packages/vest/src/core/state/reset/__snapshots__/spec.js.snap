// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vest.reset sanity 1`] = `
Object {
  "errorCount": 0,
  "getErrors": [Function],
  "getErrorsByGroup": [Function],
  "getWarnings": [Function],
  "getWarningsByGroup": [Function],
  "groups": Object {},
  "hasErrors": [Function],
  "hasErrorsByGroup": [Function],
  "hasWarnings": [Function],
  "hasWarningsByGroup": [Function],
  "name": "suite",
  "testCount": 0,
  "tests": Object {},
  "warnCount": 0,
}
`;

exports[`vest.reset sanity 2`] = `
Object {
  "errorCount": 2,
  "getErrors": [Function],
  "getErrorsByGroup": [Function],
  "getWarnings": [Function],
  "getWarningsByGroup": [Function],
  "groups": Object {},
  "hasErrors": [Function],
  "hasErrorsByGroup": [Function],
  "hasWarnings": [Function],
  "hasWarningsByGroup": [Function],
  "name": "suite",
  "testCount": 4,
  "tests": Object {
    "f1": Object {
      "errorCount": 1,
      "errors": Array [
        "m1",
      ],
      "testCount": 1,
      "warnCount": 0,
    },
    "f2": Object {
      "errorCount": 1,
      "errors": Array [
        "m2",
      ],
      "testCount": 2,
      "warnCount": 1,
      "warnings": Array [
        "m2",
      ],
    },
    "f3": Object {
      "errorCount": 0,
      "testCount": 1,
      "warnCount": 0,
    },
  },
  "warnCount": 1,
}
`;
